homeassistant/components/ihc/__init__.py

```diff
@@ -0,0 +1,122 @@
+        ihcid = int(call.data.get('ihcid', 0))
+        value = bool(call.data.get('value', 0))
+        ihc.set_runtime_value_bool(ihcid, value)
+
+    def set_runtime_value_int(call):
+        """Set a IHC runtime integer value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = int(call.data.get('value', 0))
+        ihc.set_runtime_value_int(ihcid, value)
+
+    def set_runtime_value_float(call):
+        """Set a IHC runtime float value service function."""
+        ihcid = int(call.data.get('ihcid', 0))
+        value = float(call.data.get('value', 0))
+        ihc.set_runtime_value_float(ihcid, value)
+
+    descriptions = load_yaml_config_file(
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_BOOL,
+                           set_runtime_value_bool,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_BOOL])
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_INT,
+                           set_runtime_value_int,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_INT])
+    hass.services.register(DOMAIN, const.SERVICE_SET_RUNTIME_VALUE_FLOAT,
+                           set_runtime_value_float,
+                           descriptions[const.SERVICE_SET_RUNTIME_VALUE_FLOAT])
+    return True
+
+
+class IHCPlatform:
+    """Wraps the IHCController for caching of ihc project and autosetup."""
+
+    def __init__(self, ihccontroller):
+        """Initialize the caching of the project."""
+        self.ihc = ihccontroller
+        project = self.ihc.get_project()
+        self._project = xml.etree.ElementTree.fromstring(project)
+        self._groups = self._project.findall(r'.//group')
+
+    def get_project_xml(self):
+        """Get the cached ihc project as xml."""
+        return self._project
+
+    def get_groups_xml(self):
+        """Get the groups from the ihc project and cache the result."""
+        return self._groups
+
+    def autosetup(self, productautosetup, callback):
+        """Do autosetup of a component usign the specified productautosetup."""
+        groups = self.get_groups_xml()
+        for group in groups:
+            groupname = group.attrib['name']
+            for productcfg in productautosetup:
+                products = group.findall(productcfg['xpath'])
+                for product in products:
+                    nodes = product.findall(productcfg['node'])
+                    for node in nodes:
+                        if ('setting' in node.attrib
+                                and node.attrib['setting'] == 'yes'):
+                            continue
+                        ihcid = int(node.attrib['id'].strip('_'), 0)
+                        name = groupname + "_" + str(ihcid)
+                        callback(ihcid, name, product, productcfg)
+
+
+def get_ihc_platform(hass) -> IHCPlatform:
```

Comment:
I have removed it. I think this was because I was missing the dependency to the Ihc component, so sometimes a platform would load before the ihc component.